common/content/statusline.js

```diff
@@ -145,40 +141,62 @@ const StatusLine = Module("statusline", {
 
         if (!bb) return;
 
-        var toggle_off = function () {
+        var hideStatusline = function () {
+            // Do nothing if statusline is invisible, because it would store an invalid version of ContentSeparator.
+            // Do nothing if we are in commandline mode, because the user interacts with the statusline.
+            if (!sv.isVisible || liberator.mode == modes.COMMAND_LINE) {
+                return;
+            }
+
             bb.style.height = '0px';
             bb.style.overflow = 'hidden';
+            sv.contentSeparator = highlight.get('ContentSeparator').value;
             highlight.set('ContentSeparator', 'display: none;');
+            sv.isVisible = false;
         };
 
-        var toggle_on = function () {
+        var showStatusline = function () {
+            if (sv.isVisible) {
+                return;
+            }
+
             bb.style.height = '';
             bb.style.overflow = '';
             highlight.set('ContentSeparator', sv.contentSeparator);
+            sv.isVisible = true;
         };
 
+        let mode = options["statuslinevisibility"];
+
         switch (request) {
-        case sv.TOGGLE:
-            sv.currVisibility = !sv.currVisibility;
-            if (sv.currVisibility) toggle_on();
-            else toggle_off();
-            break;
-        case sv.FULLSCREEN:
-            if (window.fullScreen) {
-                sv.prevVisibility = sv.currVisibility;
-                sv.currVisibility = false;
-                toggle_off();
-            } else {
-                sv.currVisibility = sv.currVisibility || sv.prevVisibility;
-                if (sv.currVisibility) toggle_on();
-            }
-            break;
-        case sv.MODE_ON:
-            if (!sv.currVisibility) toggle_on();
-            break;
-        case sv.MODE_OFF:
-            if (!sv.currVisibility) toggle_off();
-            break;
+            case sv.UPDATE:
+                switch (mode) {
+                    case "auto":
+                        if (window.fullScreen) {
+                            hideStatusline();
+                        } else {
+                            showStatusline();
+                        }
+                        break;
+                    case "visible":
+                        showStatusline();
+                        break;
+                    case "hidden":
+                        hideStatusline();
+                        break;
+                }
+                break;
+
+            case sv.SHOW:
+                showStatusline();
+                break;
+
+            case sv.HIDE:
+                // Only hide when in auto+fullscreen or hidden.
+                if ((mode == "auto" && window.fullScreen) || mode == "hidden") {
```

Comment:
I'm not sure that `window.fullScreen` is what we should be using here. It works which is good but it is non-standard and we should try and avoid that as much as possible.
The vendor-prefixed native fullscreen stuff, however, doesn't help us when you put the browser into fullscreen directly via something like `<F11>`. Trying to get help in mozilla's #extdev irc channel, will report back any findings.